File: backend/app/schemas/workflow.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field
from app.schemas.common import CamelModel


class WorkflowAction(CamelModel):
    action: str
    next_state: str | None = None
    allowed: bool = True
    label: str | None = None


class WorkflowDocumentContext(CamelModel):
    doctype: str
    name: str
    workflow_state: str | None = None
    available_actions: list[WorkflowAction] = Field(default_factory=list)
    source_message_id: str | None = None
    source_result_id: str | None = None


class WorkflowActionPreviewRequest(CamelModel):
    doctype: str
    name: str
    action: str
    comment: str | None = None


class WorkflowActionPreviewResponse(CamelModel):
    doctype: str
    name: str
    action: str
    current_state: str | None = None
    next_state: str | None = None
    title: str | None = None
    summary: dict[str, Any] = Field(default_factory=dict)
    confirmation_required: bool = True
    confirmation_id: str


class PendingWorkflowDocument(CamelModel):
    doctype: str
    name: str
    title: str | None = None
    workflow_state: str | None = None
    status: str | None = None
    owner: str | None = None
    modified: str | None = None
    posting_date: str | None = None
    transaction_date: str | None = None
    party: str | None = None
    grand_total: float | None = None
    currency: str | None = None
    available_actions: list[WorkflowAction] = Field(default_factory=list)


class PendingApprovalsResponse(CamelModel):
    documents: list[PendingWorkflowDocument]
    total: int
    filters: dict[str, Any] = Field(default_factory=dict)


class WorkflowDocumentDetail(CamelModel):
    doctype: str
    name: str
    title: str | None = None
    workflow_state: str | None = None
    status: str | None = None
    docstatus: int | None = None
    summary: dict[str, Any] = Field(default_factory=dict)
    fields: dict[str, Any] = Field(default_factory=dict)
    items: list[dict[str, Any]] = Field(default_factory=list)
    available_actions: list[WorkflowAction] = Field(default_factory=list)
    permission: dict[str, Any] | None = None


class ApplyWorkflowActionRequest(CamelModel):
    doctype: str
    name: str
    action: str
    comment: str | None = None
    confirmation_id: str


class ApplyWorkflowActionResponse(CamelModel):
    doctype: str
    name: str
    action: str
    previous_state: str | None = None
    new_state: str | None = None
    status: str | None = None
    available_actions: list[WorkflowAction] = Field(default_factory=list)
    message: str
    result: dict[str, Any] = Field(default_factory=dict)

File: src/types/workflow.ts
export type WorkflowAction = {
  action: string
  nextState?: string | null
  next_state?: string | null
  allowed?: boolean
  label?: string | null
}

export type PendingWorkflowDocument = {
  doctype: string
  name: string
  title?: string | null
  workflowState?: string | null
  workflow_state?: string | null
  status?: string | null
  owner?: string | null
  modified?: string | null
  postingDate?: string | null
  posting_date?: string | null
  transactionDate?: string | null
  transaction_date?: string | null
  party?: string | null
  grandTotal?: number | null
  grand_total?: number | null
  currency?: string | null
  availableActions?: WorkflowAction[]
  available_actions?: WorkflowAction[]
}

export type PendingApprovalsResponse = {
  documents: PendingWorkflowDocument[]
  total: number
  filters?: Record<string, unknown>
}

export type WorkflowDocumentDetail = {
  doctype: string
  name: string
  title?: string | null
  workflowState?: string | null
  workflow_state?: string | null
  status?: string | null
  docstatus?: number | null
  summary: Record<string, unknown>
  fields: Record<string, unknown>
  items: Array<Record<string, unknown>>
  availableActions?: WorkflowAction[]
  available_actions?: WorkflowAction[]
  permission?: Record<string, unknown> | null
}

export type ApplyWorkflowActionRequest = {
  doctype: string
  name: string
  action: string
  comment?: string
  confirmationId?: string
  confirmation_id?: string
}

export type ApplyWorkflowActionResponse = {
  doctype: string
  name: string
  action: string
  previousState?: string | null
  previous_state?: string | null
  newState?: string | null
  new_state?: string | null
  status?: string | null
  availableActions?: WorkflowAction[]
  available_actions?: WorkflowAction[]
  message: string
  result?: Record<string, unknown>
}

export type WorkflowActionPreviewRequest = {
  doctype: string
  name: string
  action: string
  comment?: string
}

export type WorkflowActionPreviewResponse = {
  doctype: string
  name: string
  action: string
  currentState?: string | null
  current_state?: string | null
  nextState?: string | null
  next_state?: string | null
  title?: string | null
  summary?: Record<string, unknown>
  confirmationRequired?: boolean
  confirmation_required?: boolean
  confirmationId?: string
  confirmation_id?: string
}
